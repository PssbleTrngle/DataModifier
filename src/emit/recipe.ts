import { Acceptor, exists } from '@pssbletrngle/pack-resolver'
import { encodeId, Id, IdInput, NormalizedId } from '../common/id.js'
import {
   CommonTest,
   createIngredient,
   Ingredient,
   IngredientInput,
   IngredientTest,
   Predicate,
   resolveIngredientTest,
} from '../common/ingredient.js'
import { resolveIDTest } from '../common/predicates.js'
import { createResult, Result, ResultInput } from '../common/result.js'
import RegistryLookup from '../loader/registry/index.js'
import { TagRegistryHolder } from '../loader/tags.js'
import { Logger } from '../logger.js'
import { createReplacer, Recipe, Replacer } from '../parser/recipe/index.js'
import { Modifier } from '../rule/index.js'
import RecipeRule from '../rule/recipe.js'
import { RecipeDefinition } from '../schema/recipe.js'
import CustomEmitter from './custom.js'
import { RegistryProvider } from './index.js'
import RuledEmitter from './ruled.js'
import { RecipeSerializerId } from '@pssbletrngle/data-modifier/generated'

export type RecipeTest = Readonly<{
   id?: CommonTest<NormalizedId>
   type?: CommonTest<NormalizedId<RecipeSerializerId>>
   namespace?: string
   output?: IngredientTest
   input?: IngredientTest
   optional?: boolean
}>

export interface RecipeRules {
   replaceResult(test: IngredientTest, value: Result, additionalTests?: RecipeTest): void

   replaceIngredient(test: IngredientTest, value: Ingredient, additionalTests?: RecipeTest): void

   add<TDefinition extends RecipeDefinition, TRecipe extends Recipe<TDefinition>>(
      id: IdInput,
      value: TDefinition | TRecipe
   ): void

   remove(test: RecipeTest): void
}

export const EMPTY_RECIPE: RecipeDefinition = {
   type: 'noop',
   conditions: [
      {
         type: 'forge:false',
      },
   ],
   'fabric:load_conditions': [
      {
         condition: 'fabric:not',
         value: {
            condition: 'fabric:all_mods_loaded',
            values: ['minecraft'],
         },
      },
   ],
}

export default class RecipeEmitter implements RecipeRules {
   private readonly custom = new CustomEmitter<RecipeDefinition>(this.recipePath)

   private readonly ruled = new RuledEmitter<Recipe, RecipeRule>(
      this.logger,
      this.registry,
      this.recipePath,
      EMPTY_RECIPE,
      id => this.custom.has(id)
   )

   constructor(
      private readonly logger: Logger,
      private readonly registry: RegistryProvider<Recipe>,
      private readonly tags: TagRegistryHolder,
      private readonly lookup: () => RegistryLookup
   ) {}

   private recipePath(id: Id) {
      return `data/${id.namespace}/recipes/${id.path}.json`
   }

   async emit(acceptor: Acceptor) {
      await Promise.all([this.ruled.emit(acceptor), this.custom.emit(acceptor)])
   }

   resolveIngredientTest(test?: IngredientTest) {
      if (!test) return () => true
      return resolveIngredientTest(test, this.tags, this.lookup())
   }

   private resolveRecipeTest(test: RecipeTest) {
      const id: Predicate<Id>[] = []
      const type: Predicate<Id>[] = []
      const ingredient: Predicate<IngredientInput>[] = []
      const result: Predicate<IngredientInput>[] = []

      if (test.id) id.push(resolveIDTest(test.id))
      if (test.type) type.push(resolveIDTest(test.type))
      if (test.namespace) id.push(id => id.namespace === test.namespace)
      if (test.output) result.push(this.resolveIngredientTest(test.output))
      if (test.input) ingredient.push(this.resolveIngredientTest(test.input))

      return { id, type, ingredient, result }
   }

   add<TDefinition extends RecipeDefinition, TRecipe extends Recipe<TDefinition>>(
      id: IdInput,
      value: TDefinition | TRecipe
   ) {
      if (this.custom.has(id)) this.logger.error(`Overwriting custom recipe with ID ${encodeId(id)}`)
      if (value instanceof Recipe) this.custom.add(id, value.toJSON())
      else this.custom.add(id, value)
   }

   private addRule(
      shape: unknown[],
      modifier: Modifier<Recipe>,
      recipeTest: RecipeTest = {},
      ingredientTests: { ingredient?: Predicate<IngredientInput>; result?: Predicate<IngredientInput> } = {}
   ) {
      const recipePredicates = this.resolveRecipeTest(recipeTest ?? {})

      this.ruled.addRule(
         new RecipeRule(
            shape,
            recipePredicates.id,
            recipePredicates.type,
            [ingredientTests.ingredient, ...recipePredicates.ingredient].filter(exists),
            [ingredientTests.result, ...recipePredicates.result].filter(exists),
            modifier
         ),
         recipeTest.optional !== true
      )
   }

   remove(test: RecipeTest) {
      this.addRule([test], () => null, test)
   }

   replaceResult(test: IngredientTest, value: ResultInput, additionalTest?: RecipeTest) {
      const predicate = this.resolveIngredientTest(test)

      createResult(value, this.lookup())
      const replacer = createReplacer<ResultInput>(predicate, value)
      const replace: Replacer<Result> = it => createResult(replacer(it))

      this.addRule(
         ['replace result', test, 'with', value, additionalTest],
         recipe => recipe.replaceResult(replace),
         additionalTest,
         { result: predicate }
      )
   }

   replaceIngredient(test: IngredientTest, value: IngredientInput, additionalTest?: RecipeTest) {
      const predicate = this.resolveIngredientTest(test)

      createIngredient(value, this.lookup())
      const replacer = createReplacer<IngredientInput>(predicate, value)
      const replace: Replacer<Ingredient> = it => createIngredient(replacer(it))

      this.addRule(
         ['replace ingredient', test, 'with', value, additionalTest],
         recipe => recipe.replaceIngredient(replace),
         additionalTest,
         { ingredient: predicate }
      )
   }

   clear() {
      this.custom.clear()
      this.ruled.clear()
   }
}
