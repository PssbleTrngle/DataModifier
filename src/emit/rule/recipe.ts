import { exists } from '@pssbletrngle/pack-resolver'
import { Id, createId } from '../../common/id.js'
import { IngredientInput, Predicate } from '../../common/ingredient.js'
import { Logger } from '../../logger.js'
import { Recipe } from '../../parser/recipe/index.js'
import Rule, { Modifier } from './index.js'

export default class RecipeRule extends Rule<Recipe> {
   constructor(
      private readonly shape: unknown[],
      private readonly idsTests: Predicate<Id>[],
      private readonly typeTests: Predicate<Id>[],
      private readonly ingredientTests: Predicate<IngredientInput>[],
      private readonly resultTests: Predicate<IngredientInput>[],
      modifier: Modifier<Recipe>
   ) {
      super(modifier)
   }

   matches(id: Id, recipe: Recipe, logger: Logger): boolean {
      const types = recipe.getTypes().map(createId)
      return (
         this.idsTests.every(test => test(id, logger)) &&
         this.typeTests.every(test => types.some(it => test(it))) &&
         this.ingredientTests.every(test => recipe.getIngredients().some(it => test(it, logger))) &&
         this.resultTests.every(test => recipe.getResults().some(it => test(it, logger)))
      )
   }

   printWarning(logger: Logger) {
      logger.error('Could not find any recipes matching', ...this.shape.filter(exists))
   }
}
