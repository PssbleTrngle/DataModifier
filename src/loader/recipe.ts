import RecipeParser, { Recipe } from '../parser/recipe/index.js'
import ShapedParser from '../parser/recipe/vanilla/shaped.js'
import { RecipeDefinition } from '../schema/data/recipe.js'
import { encodeId } from '../common/id.js'
import ShapelessParser from '../parser/recipe/vanilla/shapeless.js'
import CreateProcessingRecipeParser from '../parser/recipe/create/processing.js'
import SmeltingParser from '../parser/recipe/vanilla/smelting.js'
import SmithingParser from '../parser/recipe/vanilla/smithing.js'
import AssemblyRecipeParser from '../parser/recipe/create/assembly.js'
import StonecuttingParser from '../parser/recipe/vanilla/stonecutting.js'
import CuttingRecipeParser from '../parser/recipe/farmersdelight/cutting.js'
import CookingRecipeParser from '../parser/recipe/farmersdelight/cooking.js'
import ThermalRecipeParser from '../parser/recipe/thermal/index.js'
import ThermalCatalystRecipeParser from '../parser/recipe/thermal/catalyst.js'
import NbtWrapperRecipeParser from '../parser/recipe/botania/nbtWrapper.js'
import OrechidRecipeParser from '../parser/recipe/botania/orechid.js'
import RunicAltarRecipeParser from '../parser/recipe/botania/runicAltar.js'
import ElvenTradeRecipeParser from '../parser/recipe/botania/elvenTrade.js'
import BrewRecipeParser from '../parser/recipe/botania/brew.js'
import ManaInfusionRecipeParser from '../parser/recipe/botania/manaInfusion.js'
import GogWrapperRecipeParser from '../parser/recipe/botania/gogWrapper.js'
import ApothecaryRecipeParser from '../parser/recipe/botania/apothecary.js'
import TerraPlateRecipeParser from '../parser/recipe/botania/terraPlate.js'
import PureDaisyRecipeParser from '../parser/recipe/botania/pureDaisy.js'
import { JsonLoader } from './index.js'
import ForgeConditionalRecipeParser from '../parser/recipe/forge/conditional.js'
import HammeringRecipeParser from '../parser/recipe/adAstra/hammering.js'
import InputOutputRecipeParser from '../parser/recipe/adAstra/inputOutput.js'
import NasaWorkbenchRecipeParser from '../parser/recipe/adAstra/nasaWorkbench.js'
import SpaceStationRecipeParser from '../parser/recipe/adAstra/spaceStation.js'
import QuarkExclusionRecipeParser from '../parser/recipe/quark/exclusion.js'
import TreeExtractionRecipeParser from '../parser/recipe/thermal/treeExtraction.js'
import ThermalFuelRecipeParser from '../parser/recipe/thermal/fuel.js'
import RootComponentRecipeParser from '../parser/recipe/roots/component.js'
import RootRitualRecipeParser from '../parser/recipe/roots/ritual.js'
import { ShapelessRecipeParser } from '../parser/index.js'
import { Logger } from '../logger.js'
import { IllegalShapeError } from '../error.js'
import IgnoredRecipe from '../parser/recipe/ignored.js'
import { createResult } from '../common/result.js'
import { createIngredient } from '../common/ingredient.js'
import FluidConversionRecipeParser from '../parser/recipe/adAstra/conversion.js'

export interface RecipeLoaderAccessor {
   unknownRecipeTypes(): RecipeDefinition[]
   registerParser(recipeType: string, parser: RecipeParser<RecipeDefinition, Recipe>): void
   ignoreType(recipeType: string): void
}

export default class RecipeLoader extends JsonLoader<Recipe> implements RecipeLoaderAccessor {
   private readonly recipeParsers = new Map<string, RecipeParser<RecipeDefinition, Recipe>>()

   private readonly ignoredRecipeTypes = new Set<string>()
   private readonly _unknownRecipeTypes = new Map<string, RecipeDefinition>()

   constructor() {
      super()

      this.registerParser('minecraft:crafting_shaped', new ShapedParser())
      this.registerParser('minecraft:crafting_shapeless', new ShapelessParser())
      this.registerParser('minecraft:smelting', new SmeltingParser())
      this.registerParser('minecraft:smoking', new SmeltingParser())
      this.registerParser('minecraft:blasting', new SmeltingParser())
      this.registerParser('minecraft:campfire_cooking', new SmeltingParser())
      this.registerParser('minecraft:smithing', new SmithingParser())
      this.registerParser('minecraft:stonecutting', new StonecuttingParser())

      this.registerParser('create:mixing', new CreateProcessingRecipeParser())
      this.registerParser('create:pressing', new CreateProcessingRecipeParser())
      this.registerParser('create:emptying', new CreateProcessingRecipeParser())
      this.registerParser('create:crushing', new CreateProcessingRecipeParser())
      this.registerParser('create:milling', new CreateProcessingRecipeParser())
      this.registerParser('create:compacting', new CreateProcessingRecipeParser())
      this.registerParser('create:filling', new CreateProcessingRecipeParser())
      this.registerParser('create:cutting', new CreateProcessingRecipeParser())
      this.registerParser('create:item_application', new CreateProcessingRecipeParser())
      this.registerParser('create:sandpaper_polishing', new CreateProcessingRecipeParser())
      this.registerParser('create:deploying', new CreateProcessingRecipeParser())
      this.registerParser('create:splashing', new CreateProcessingRecipeParser())
      this.registerParser('create:haunting', new CreateProcessingRecipeParser())
      this.registerParser('create:mechanical_crafting', new ShapedParser())
      this.registerParser('create:sequenced_assembly', new AssemblyRecipeParser())

      this.registerParser('farmersdelight:cooking', new CookingRecipeParser())
      this.registerParser('farmersdelight:cutting', new CuttingRecipeParser())
      this.registerParser('farmersrespite:brewing', new CookingRecipeParser())

      this.registerParser('thermal:bottler', new ThermalRecipeParser())
      this.registerParser('thermal:centrifuge', new ThermalRecipeParser())
      this.registerParser('thermal:chiller', new ThermalRecipeParser())
      this.registerParser('thermal:crucible', new ThermalRecipeParser())
      this.registerParser('thermal:crystallizer', new ThermalRecipeParser())
      this.registerParser('thermal:furnace', new ThermalRecipeParser())
      this.registerParser('thermal:insolator', new ThermalRecipeParser())
      this.registerParser('thermal:insolator_catalyst', new ThermalCatalystRecipeParser())
      this.registerParser('thermal:press', new ThermalRecipeParser())
      this.registerParser('thermal:pulverizer', new ThermalRecipeParser())
      this.registerParser('thermal:pulverizer_recycle', new ThermalRecipeParser())
      this.registerParser('thermal:pulverizer_catalyst', new ThermalCatalystRecipeParser())
      this.registerParser('thermal:pyrolyzer', new ThermalRecipeParser())
      this.registerParser('thermal:refinery', new ThermalRecipeParser())
      this.registerParser('thermal:sawmill', new ThermalRecipeParser())
      this.registerParser('thermal:smelter', new ThermalRecipeParser())
      this.registerParser('thermal:smelter_recycle', new ThermalRecipeParser())
      this.registerParser('thermal:smelter_catalyst', new ThermalCatalystRecipeParser())
      this.registerParser('thermal:tree_extractor', new TreeExtractionRecipeParser())
      this.registerParser('thermal:compression_fuel', new ThermalFuelRecipeParser())
      this.registerParser('thermal:magmatic_fuel', new ThermalFuelRecipeParser())
      this.registerParser('thermal:gourmand_fuel', new ThermalFuelRecipeParser())
      this.registerParser('thermal:numismatic_fuel', new ThermalFuelRecipeParser())

      this.registerParser('botania:nbt_output_wrapper', new NbtWrapperRecipeParser())
      this.registerParser('botania:orechid', new OrechidRecipeParser())
      this.registerParser('botania:orechid_ignem', new OrechidRecipeParser())
      this.registerParser('botania:marimorphosis', new OrechidRecipeParser())
      this.registerParser('botania:pure_daisy', new PureDaisyRecipeParser())
      this.registerParser('botania:state_copying_pure_daisy', new PureDaisyRecipeParser())
      this.registerParser('botania:mana_upgrade', new ShapedParser())
      this.registerParser('botania:water_bottle_matching_shaped', new ShapedParser())
      this.registerParser('botania:runic_altar', new RunicAltarRecipeParser())
      this.registerParser('botania:runic_altar_head', new RunicAltarRecipeParser())
      this.registerParser('botania:terra_plate', new TerraPlateRecipeParser())
      this.registerParser('botania:elven_trade', new ElvenTradeRecipeParser())
      this.registerParser('botania:brew', new BrewRecipeParser())
      this.registerParser('botania:twig_wand', new ShapedParser())
      this.registerParser('botania:mana_infusion', new ManaInfusionRecipeParser())
      this.registerParser('botania:mana_upgrade_shapeless', new ShapelessParser())
      this.registerParser('botania:armor_upgrade', new ShapedParser())
      this.registerParser('botania:gog_alternation', new GogWrapperRecipeParser())
      this.registerParser('botania:petal_apothecary', new ApothecaryRecipeParser())

      this.registerParser('theoneprobe:probe_helmet', new ShapedParser())

      this.registerParser('forge:conditional', new ForgeConditionalRecipeParser())

      this.registerParser('sullysmod:grindstone_polishing', new SmeltingParser())

      this.registerParser('ad_astra:hammering', new HammeringRecipeParser())
      this.registerParser('ad_astra:cryo_fuel_conversion', new FluidConversionRecipeParser())
      this.registerParser('ad_astra:fuel_conversion', new FluidConversionRecipeParser())
      this.registerParser('ad_astra:oxygen_conversion', new FluidConversionRecipeParser())
      this.registerParser('ad_astra:compressing', new InputOutputRecipeParser())
      this.registerParser('ad_astra:crafting_shaped_space_suit', new ShapedParser())
      this.registerParser('ad_astra:nasa_workbench', new NasaWorkbenchRecipeParser())
      this.registerParser('ad_astra:space_station', new SpaceStationRecipeParser())

      this.registerParser('quark:exclusion', new QuarkExclusionRecipeParser())
      this.registerParser('patchouli:shapeless_book_recipe', new ShapelessRecipeParser())

      this.registerParser('cofh_core:crafting_shaped_potion', new ShapedParser())

      this.registerParser('rootsclassic:component', new RootComponentRecipeParser())
      this.registerParser('rootsclassic:ritual', new RootRitualRecipeParser())

      this.ignoreType('jeed:effect_provider')
      this.ignoreType('jeed:potion_provider')

      this.ignoreType('pipez:copy_nbt')
      this.ignoreType('pipez:clear_nbt')

      this.ignoreType('refinedstorage:upgrade_with_enchanted_book')

      this.ignoreType('forge:ore_shaped')

      this.ignoreType('supplementaries:trapped_present')
      this.ignoreType('supplementaries:weathered_map')
      this.ignoreType('supplementaries:soap_clearing')
      this.ignoreType('supplementaries:rope_arrow_create')
      this.ignoreType('supplementaries:present_dye')
      this.ignoreType('supplementaries:rope_arrow_add')
      this.ignoreType('supplementaries:item_lore')
      this.ignoreType('supplementaries:flag_from_banner')
      this.ignoreType('supplementaries:bubble_blower_charge')
      this.ignoreType('supplementaries:bamboo_spikes_tipped')
      this.ignoreType('supplementaries:antique_book')
      this.ignoreType('supplementaries:cauldron_flag_dye')
      this.ignoreType('supplementaries:cauldron_flag_clear')
      this.ignoreType('supplementaries:cauldron_blackboard')

      this.ignoreType('quark:mixed_exclusion')
      this.ignoreType('quark:elytra_duplication')
      this.ignoreType('quark:slab_to_block')

      // TODO could to in the future
      this.ignoreType('ad_astra:lunarian_trade_simple')
      this.ignoreType('ad_astra:lunarian_trade_enchanted_item')
      this.ignoreType('ad_astra:lunarian_trade_suspicious_stew')
      this.ignoreType('ad_astra:lunarian_trade_enchanted_book')
      this.ignoreType('ad_astra:lunarian_trade_dyed_item')
      this.ignoreType('ad_astra:lunarian_trade_potioned_item')

      this.ignoreType('immersiveengineering:cloche')
      this.ignoreType('immersiveengineering:crusher')
      this.ignoreType('immersiveengineering:fermenter')
      this.ignoreType('immersiveengineering:metal_press')
      this.ignoreType('immersiveengineering:squeezer')
      this.ignoreType('immersiveengineering:mineral_mix')
   }

   ignoreType(recipeType: string) {
      this.ignoredRecipeTypes.add(recipeType)
   }

   unknownRecipeTypes() {
      return [...this._unknownRecipeTypes.values()]
   }

   parse<TDefinition extends RecipeDefinition, TRecipe extends Recipe<TDefinition>>(
      logger: Logger,
      definition: TDefinition
   ): TRecipe | null {
      const parser = this.recipeParsers.get(encodeId(definition.type))

      if (!('type' in definition)) throw new IllegalShapeError('recipe type missing')
      if (Object.keys(definition).length <= 1) return null
      if (this.ignoredRecipeTypes.has(definition.type)) return null

      if (!parser) {
         if (!this._unknownRecipeTypes.has(definition.type)) {
            this._unknownRecipeTypes.set(definition.type, definition)
            throw new IllegalShapeError(`unknown recipe type: '${definition.type}'`, definition)
         }
         return null
      }

      const parsed = parser.create(definition, it => {
         const parsed = this.parse(logger, it)
         if (parsed) return parsed
         return new IgnoredRecipe(it)
      }) as TRecipe

      // Catch warnings early
      parsed.getResults().forEach(it => createResult(it))
      parsed.getIngredients().forEach(it => createIngredient(it))

      return parsed
   }

   registerParser(recipeType: string, parser: RecipeParser<RecipeDefinition, Recipe>) {
      this.recipeParsers.set(recipeType, parser)
   }

   clear() {
      this._unknownRecipeTypes.clear()
   }
}
