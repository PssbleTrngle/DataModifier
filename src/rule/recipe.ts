import { IngredientInput, Predicate } from '../common/ingredient'
import { Recipe } from '../parser/recipe'
import { encodeId, Id } from '../common/id'
import { Logger } from '../logger'
import Rule, { Modifier } from './index'

export default class RecipeRule extends Rule<Recipe> {
   constructor(
      private readonly idsTests: Predicate<Id>[],
      private readonly ingredientTests: Predicate<IngredientInput>[],
      private readonly resultTests: Predicate<IngredientInput>[],
      modifier: Modifier<Recipe>
   ) {
      super(modifier)
   }

   matches(id: Id, recipe: Recipe, logger: Logger): boolean {
      const prefixed = logger.group(encodeId(id))
      return (
         this.idsTests.every(test => test(id, prefixed)) &&
         this.ingredientTests.every(test => recipe.getIngredients().some(it => test(it, prefixed))) &&
         this.resultTests.every(test => recipe.getResults().some(it => test(it, prefixed)))
      )
   }
}
