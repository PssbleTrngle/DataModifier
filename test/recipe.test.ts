import { describe, expect, it } from 'vitest'
import { EMPTY_RECIPE, RecipeTest } from '../src/emit/data/recipe.js'
import { NormalizedId } from '../src/index.js'
import { ShapedRecipeDefinition } from '../src/parser/recipe/vanilla/shaped.js'
import createTestAcceptor from './mock/TestAcceptor.js'
import setupLoader from './shared/loaderSetup.js'

const { logger, loader } = setupLoader({ include: ['data/**/*.json'] })

it('has no unknown recipe loaders', () => {
   expect(loader.recipeLoader.unknownRecipeTypes()).toMatchObject([])
})

it('does not encounter any errors', () => {
   expect(logger.warn).not.toHaveBeenCalled()
   expect(logger.error).not.toHaveBeenCalled()
})

describe('recipe ingredient replacement', () => {
   it('replaces ingredients', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.replaceIngredient('minecraft:redstone', {
         item: 'minecraft:emerald',
      })

      await loader.emit(acceptor)

      expect(acceptor.jsonAt('data/minecraft/recipes/piston.json')).toMatchSnapshot('modified piston recipe')
      expect(acceptor.jsonAt('data/minecraft/recipes/compass.json')).toMatchSnapshot('modified compass recipe')

      expect(acceptor.paths()).toMatchSnapshot('recipes including redstone as an ingredient')
   })

   it('replaces ingredients with additional input filter', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.replaceIngredient(
         'minecraft:redstone',
         {
            item: 'minecraft:emerald',
         },
         {
            input: '#minecraft:planks',
         }
      )

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(2)

      expect(acceptor.jsonAt('data/minecraft/recipes/piston.json')).toMatchSnapshot('modified piston recipe')
      expect(acceptor.jsonAt('data/minecraft/recipes/note_block.json')).toMatchSnapshot('modified note_block recipe')
      expect(acceptor.jsonAt('data/minecraft/recipes/compass.json')).toBeNull()
   })

   it('replaces ingredients in create recipes', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.replaceIngredient('#forge:raw_materials/zinc', {
         tag: 'forge:raw_materials/iron',
      })

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(4)

      expect(acceptor.jsonAt('data/create/recipes/crafting/materials/raw_zinc_block.json')).toMatchSnapshot(
         'modified create:raw_zinc_block recipe'
      )
      expect(acceptor.jsonAt('data/create/recipes/crushing/raw_zinc.json')).toMatchSnapshot(
         'modified create:raw_zinc recipe'
      )
      expect(acceptor.jsonAt('data/create/recipes/blasting/zinc_ingot_from_raw_ore.json')).toMatchSnapshot(
         'modified create:zinc_ingot_from_raw_ore recipe'
      )
      expect(acceptor.jsonAt('data/create/recipes/smelting/zinc_ingot_from_raw_ore.json')).toMatchSnapshot(
         'modified create:zinc_ingot_from_raw_ore recipe'
      )
   })

   it('matches recipes wrapped in forge:conditional', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.remove({ input: { item: 'biomesoplenty:violet' }, type: 'farmersdelight:cutting' })

      await loader.emit(acceptor)

      expect(acceptor.jsonAt('data/custom/recipes/conditional.json')).toMatchObject(EMPTY_RECIPE)
   })
})

describe('recipe removal', () => {
   it('removes recipes with id filter', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.remove({
         id: /minecraft:.*piston/,
      })

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(2)

      expect(acceptor.jsonAt('data/minecraft/recipes/piston.json')).toMatchObject(EMPTY_RECIPE)
      expect(acceptor.jsonAt('data/minecraft/recipes/sticky_piston.json')).toMatchObject(EMPTY_RECIPE)
   })

   it('removes recipes with type filter', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.remove({
         type: 'minecraft:smelting',
      })

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(118)
   })

   it('removes recipes with result filter', async () => {
      const acceptor = createTestAcceptor()

      loader.recipes.remove({
         output: 'minecraft:cooked_beef',
      })

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(3)

      expect(acceptor.jsonAt('data/minecraft/recipes/cooked_beef.json')).toMatchObject(EMPTY_RECIPE)
      expect(acceptor.jsonAt('data/minecraft/recipes/cooked_beef_from_smoking.json')).toMatchObject(EMPTY_RECIPE)
      expect(acceptor.jsonAt('data/minecraft/recipes/cooked_beef_from_campfire_cooking.json')).toMatchObject(
         EMPTY_RECIPE
      )
   })
})

it('creates custom recipes', async () => {
   const acceptor = createTestAcceptor()

   const recipe: ShapedRecipeDefinition = {
      type: 'minecraft:shaped',
      key: {
         A: {
            item: 'minecraft:diamond',
         },
         B: {
            tag: 'minecraft:iron_ores',
         },
      },
      result: {
         item: 'minecraft:command_block',
      },
      pattern: ['A ', ' B'],
   }

   loader.recipes.add('example:custom', recipe)

   await loader.emit(acceptor)

   expect(acceptor.jsonAt('data/example/recipes/custom.json')).toMatchObject(recipe)
})

it('warns about duplicate custom recipe IDs', () => {
   const id: NormalizedId = 'example:recipe'

   loader.recipes.add(id, { type: 'example:something' })
   loader.recipes.add(id, { type: 'example:something_else' })

   expect(logger.error).toHaveBeenCalledWith(`Overwriting custom recipe with ID ${id}`)
})

it('warns about missing recipe removal matches', async () => {
   const filter: RecipeTest = {
      type: 'example:not_existing',
   }

   loader.recipes.remove(filter)

   await loader.emit(createTestAcceptor())

   expect(logger.error).toHaveBeenCalledWith('Could not find any recipes matching', filter)
})

it('warns about missing recipe replacement matches', async () => {
   const from = 'minecraft:nothing'
   const to = { item: 'minecraft:dirt' }
   loader.recipes.replaceResult(from, to)

   await loader.emit(createTestAcceptor())

   expect(logger.error).toHaveBeenCalledWith('Could not find any recipes matching', 'replace result', from, 'with', to)
})

it('does not warn about optional missing recipe matches', async () => {
   const from = 'minecraft:nothing'
   const to = { item: 'minecraft:dirt' }
   loader.recipes.replaceResult(from, to, { optional: true })

   await loader.emit(createTestAcceptor())

   expect(logger.error).not.toHaveBeenCalled()
})
