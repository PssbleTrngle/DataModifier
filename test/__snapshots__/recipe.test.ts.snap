// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replace ingredient replacement replaces ingredients in create recipes: modified create:raw_zinc recipe 1`] = `
{
  "ingredients": [
    {
      "tag": "#forge:raw_materials/iron",
    },
  ],
  "processingTime": 400,
  "results": [
    {
      "item": "create:crushed_raw_zinc",
    },
    {
      "chance": 0.75,
      "item": "create:experience_nugget",
    },
  ],
  "type": "create:crushing",
}
`;

exports[`replace ingredient replacement replaces ingredients in create recipes: modified create:raw_zinc_block recipe 1`] = `
{
  "key": {
    "C": {
      "tag": "#forge:raw_materials/iron",
    },
  },
  "pattern": [
    "CCC",
    "CCC",
    "CCC",
  ],
  "result": {
    "item": "create:raw_zinc_block",
  },
  "type": "minecraft:crafting_shaped",
}
`;

exports[`replace ingredient replacement replaces ingredients in create recipes: modified create:zinc_ingot_from_raw_ore recipe 1`] = `
{
  "cookingtime": 100,
  "experience": 0.7,
  "ingredient": {
    "tag": "#forge:raw_materials/iron",
  },
  "result": "create:zinc_ingot",
  "type": "minecraft:blasting",
}
`;

exports[`replace ingredient replacement replaces ingredients in create recipes: modified create:zinc_ingot_from_raw_ore recipe 2`] = `
{
  "cookingtime": 200,
  "experience": 0.7,
  "ingredient": {
    "tag": "#forge:raw_materials/iron",
  },
  "result": "create:zinc_ingot",
  "type": "minecraft:smelting",
}
`;

exports[`replace ingredient replacement replaces ingredients with additional input filter: modified note_block recipe 1`] = `
{
  "key": {
    "#": {
      "tag": "minecraft:planks",
    },
    "X": {
      "item": "minecraft:emerald",
    },
  },
  "pattern": [
    "###",
    "#X#",
    "###",
  ],
  "result": {
    "item": "minecraft:note_block",
  },
  "type": "minecraft:crafting_shaped",
}
`;

exports[`replace ingredient replacement replaces ingredients with additional input filter: modified piston recipe 1`] = `
{
  "key": {
    "#": {
      "item": "minecraft:cobblestone",
    },
    "R": {
      "item": "minecraft:emerald",
    },
    "T": {
      "tag": "minecraft:planks",
    },
    "X": {
      "item": "minecraft:iron_ingot",
    },
  },
  "pattern": [
    "TTT",
    "#X#",
    "#R#",
  ],
  "result": {
    "item": "minecraft:piston",
  },
  "type": "minecraft:crafting_shaped",
}
`;

exports[`replace ingredient replacement replaces ingredients: modified compass recipe 1`] = `
{
  "key": {
    "#": {
      "item": "minecraft:iron_ingot",
    },
    "X": {
      "item": "minecraft:emerald",
    },
  },
  "pattern": [
    " # ",
    "#X#",
    " # ",
  ],
  "result": {
    "item": "minecraft:compass",
  },
  "type": "minecraft:crafting_shaped",
}
`;

exports[`replace ingredient replacement replaces ingredients: modified piston recipe 1`] = `
{
  "key": {
    "#": {
      "item": "minecraft:cobblestone",
    },
    "R": {
      "item": "minecraft:emerald",
    },
    "T": {
      "tag": "minecraft:planks",
    },
    "X": {
      "item": "minecraft:iron_ingot",
    },
  },
  "pattern": [
    "TTT",
    "#X#",
    "#R#",
  ],
  "result": {
    "item": "minecraft:piston",
  },
  "type": "minecraft:crafting_shaped",
}
`;

exports[`replace ingredient replacement replaces ingredients: recipes including redstone as an ingredient 1`] = `
[
  "data/botania/recipe/brew/speed.json",
  "data/botania/recipe/brew/weak_regeneration.json",
  "data/botania/recipe/detector_light_relay.json",
  "data/botania/recipe/ender_eye_block.json",
  "data/botania/recipe/hourglass.json",
  "data/botania/recipe/lens_mine.json",
  "data/botania/recipe/mana_detector.json",
  "data/botania/recipe/mana_infusion/mana_powder_dust.json",
  "data/botania/recipe/mana_infusion/redstone_dupe.json",
  "data/botania/recipe/mana_infusion/redstone_to_glowstone_dust.json",
  "data/botania/recipe/pattern_1_1.json",
  "data/botania/recipe/pattern_1_2.json",
  "data/botania/recipe/pattern_1_3.json",
  "data/botania/recipe/pattern_2_1.json",
  "data/botania/recipe/pattern_2_2.json",
  "data/botania/recipe/pattern_2_3.json",
  "data/botania/recipe/pattern_3_1.json",
  "data/botania/recipe/pattern_3_2.json",
  "data/botania/recipe/pattern_donut.json",
  "data/botania/recipe/quartz_red.json",
  "data/botania/recipe/red_pavement.json",
  "data/botania/recipe/redstone_root.json",
  "data/botania/recipe/redstone_spreader.json",
  "data/botania/recipe/spark_changer.json",
  "data/botania/recipe/unholy_cloak.json",
  "data/minecraft/recipe/clock.json",
  "data/minecraft/recipe/compass.json",
  "data/minecraft/recipe/detector_rail.json",
  "data/minecraft/recipe/dispenser.json",
  "data/minecraft/recipe/dropper.json",
  "data/minecraft/recipe/note_block.json",
  "data/minecraft/recipe/observer.json",
  "data/minecraft/recipe/piston.json",
  "data/minecraft/recipe/powered_rail.json",
  "data/minecraft/recipe/redstone_block.json",
  "data/minecraft/recipe/redstone_lamp.json",
  "data/minecraft/recipe/redstone_torch.json",
  "data/minecraft/recipe/repeater.json",
  "data/minecraft/recipe/target.json",
]
`;
